sdk/identity/azure-identity/test/ut/token_cache_test.cpp

```diff
@@ -0,0 +1,339 @@
+  EXPECT_FALSE(key4 < key1);
+  EXPECT_FALSE(key5 < key1);
+
+  EXPECT_TRUE(key2 < key3);
+  EXPECT_TRUE(key2 < key4);
+  EXPECT_TRUE(key2 < key5);
+  EXPECT_FALSE(key3 < key2);
+  EXPECT_FALSE(key4 < key2);
+  EXPECT_FALSE(key5 < key2);
+
+  EXPECT_TRUE(key3 < key4);
+  EXPECT_TRUE(key3 < key5);
+  EXPECT_FALSE(key4 < key3);
+  EXPECT_FALSE(key5 < key3);
+
+  EXPECT_TRUE(key4 < key5);
+  EXPECT_FALSE(key5 < key4);
+}
+
+TEST(TokenCache, GetReuseRefresh)
+{
+  TokenCache::Clear();
+
+  EXPECT_EQ(TokenCache::Internals::Cache.size(), 0UL);
+
+  DateTime const Tomorrow = std::chrono::system_clock::now() + 24h;
+  auto const Yesterday = Tomorrow - 48h;
+
+  {
+    auto const token1 = TokenCache::GetToken("A", "B", "C", "D", [=]() {
+      AccessToken result;
+      result.Token = "T1";
+      result.ExpiresOn = Tomorrow;
+      return result;
+    });
+
+    EXPECT_EQ(TokenCache::Internals::Cache.size(), 1UL);
+
+    EXPECT_EQ(token1.ExpiresOn, Tomorrow);
+    EXPECT_EQ(token1.Token, "T1");
+
+    auto const token2 = TokenCache::GetToken("A", "B", "C", "D", [=]() {
+      EXPECT_FALSE("getNewToken does not get invoked when the existing cache value is good");
+      AccessToken result;
+      result.Token = "T2";
+      result.ExpiresOn = Tomorrow + 24h;
+      return result;
+    });
+
+    EXPECT_EQ(TokenCache::Internals::Cache.size(), 1UL);
+
+    EXPECT_EQ(token1.ExpiresOn, token2.ExpiresOn);
+    EXPECT_EQ(token1.Token, token2.Token);
+  }
+
+  {
+    TokenCache::Internals::Cache[{"A", "B", "C", "D"}]->AccessToken.ExpiresOn = Yesterday;
+
+    auto const token = TokenCache::GetToken("A", "B", "C", "D", [=]() {
+      AccessToken result;
+      result.Token = "T3";
+      result.ExpiresOn = Tomorrow + 1min;
+      return result;
+    });
+
+    EXPECT_EQ(TokenCache::Internals::Cache.size(), 1UL);
+
+    EXPECT_EQ(token.ExpiresOn, Tomorrow + 1min);
+    EXPECT_EQ(token.Token, "T3");
+  }
+}
+
+TEST(TokenCache, TwoThreadsAttemptToInsertTheSameKey)
+{
+  TokenCache::Clear();
+
+  EXPECT_EQ(TokenCache::Internals::Cache.size(), 0UL);
+
+  DateTime const Tomorrow = std::chrono::system_clock::now() + 24h;
+
+  TokenCache::Internals::OnBeforeCacheWriteLock = [=]() {
+    TokenCache::Internals::OnBeforeCacheWriteLock = nullptr;
+    static_cast<void>(TokenCache::GetToken("A", "B", "C", "D", [=]() {
+      AccessToken result;
+      result.Token = "T1";
+      result.ExpiresOn = Tomorrow;
+      return result;
+    }));
+  };
+
+  auto const token = TokenCache::GetToken("A", "B", "C", "D", [=]() {
+    EXPECT_FALSE("getNewToken does not get invoked when the fresh value was inserted just before "
+                 "acquiring cache write lock");
+    AccessToken result;
+    result.Token = "T2";
+    result.ExpiresOn = Tomorrow + 1min;
+    return result;
+  });
+
+  EXPECT_EQ(TokenCache::Internals::Cache.size(), 1UL);
+
+  EXPECT_EQ(token.ExpiresOn, Tomorrow);
+  EXPECT_EQ(token.Token, "T1");
+}
+
+TEST(TokenCache, TwoThreadsAttemptToUpdateTheSameToken)
+{
```

Comment:
In addition to simulating what happens when multiple threads attempt to retreive a token, you might consider a test which actually retrieves a token from a few threads at the same time.